### ISSUE TITLE
[EKS] [request]: taint ephemeral nodes that are created during Managed Node Group AMI upgrade to reduce churn

### ISSUE THREAD

**Issue by mastermanu** [aws/containers-roadmap#1384]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**
Please see attached PDF for description / images. Giving high-level overview here:

When performing an AMI update (within the same version of K8S) on our EKS Managed Node Worker group, our application (and thus, our end-users) are impacted much more heavily than we anticipate.

Because EKS relies on increasing the corresponding Auto-Scaling group's desired count to create nodes with the newer AMI version, we end up with a situation where pods from an older draining node move to a freshly created node that will then be shutdown during the scale-down phase of the AMI group at the end of the upgrade. This causes unnecessary pod churn and unnecessarily impacts availability.

**Which service(s) is this request for?**
EKS

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
Increased stability during AMI upgrades on managed node groups (see attached PDF for diagram)

**Are you currently working around this issue?**
Not really working around it for now, although creating a new node group and doing this ourselves is always a possibility.

**Attachments**
[EKS_AMI_UPGRADE.pdf](https://github.com/aws/containers-roadmap/files/6534836/EKS_AMI_UPGRADE.pdf)

**Comment by mikestef9** [comment]
Hi @mastermanu 

Do you set [pod disruption budgets](https://kubernetes.io/docs/tasks/run-application/configure-pdb/) for your applications?

**Comment by mastermanu** [comment]
> Hi @mastermanu
> 
> Do you set [pod disruption budgets](https://kubernetes.io/docs/tasks/run-application/configure-pdb/) for your applications?

Hi! Yes, we set 10% and 10% for unavailable/surge respectively and have sufficient delay between sets of pods that are upgraded.

**Comment by jaimehrubiks** [comment]
Is this "update behavior" the reason why a single-node ASG (with 3 AZs), ends up with 6 instances during the update process, just before 5 of them are terminated?

Isn't it weird to have a desired count to 1, and the managed group increasing it up to 6 just for the update? This update could be done with just 2 instances.

**Comment by michael-careplanner** [comment]
We have recently come across this behaviour when evaluating switching to managed node groups and the excessive node churn during the rolling updates is potentially going to be a blocker for us adopting them over self managed node groups.

Another idea for how to efficiently upgrade a node group would be to leverage scale in protection for instances in the ASG that is being updated to allow for upgrading while creating no unnecessary nodes that will be quickly terminated during the final scale down. It could work as follows:

 - Scale up the ASG by the max concurrency and wait for the new nodes to become ready.
 - Once all the new nodes are ready, enable scale in protection on all instance in the ASG.
 - Set the desired capacity of the ASG back to the original size. As all instances have scale in protection, the ASG will not terminate any instances automatically.
 - One by one, cordon, drain and terminate max concurrency number of outdated nodes.
 - At this point, we have a partially (or completely) upgraded node group without creating any ephemeral nodes.
 - Repeat the previous steps until there are no old instances left in the node group.
 - Remove scale in protection from all instances.

**Comment by ravisinha0506** [comment]
Hi,

EKS is exploring options to handle this scenario efficiently. We will keep you posted on this.

Thanks,

**Comment by akestner** [comment]
We recently updated the behavior of EKS Managed Nodes Groups to avoid this problem. There is no action required to benefit from these changes. @ravisinha0506 can describe these changes in more detail and then we'll close this issue.

**Comment by akestner** [comment]
We recently updated the behavior of EKS Managed Nodes Group upgrades to avoid this problem. The approach we took leverages an existing EC2 Autoscaling Group API ([`TerminateInstanceInAutoScalingGroup`](https://docs.aws.amazon.com/autoscaling/ec2/APIReference/API_TerminateInstanceInAutoScalingGroup.html)) instead of the solution described here, but has the same effect of reducing pod churn during upgrades. 

There is no additional action required to take advantage of these changes. They'll automatically be used during MNG upgrades.
